crates/stackable-operator/src/commons/networking.rs

```diff
@@ -1,44 +1,136 @@
 
 use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
+use snafu::Snafu;
 
 use crate::validation;
 
-/// A validated hostname type, for use in CRDs.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+/// A validated domain name type conforming to RFC 1123, so e.g. not an IP addresses
+#[derive(
+    Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, JsonSchema,
+)]
 #[serde(try_from = "String", into = "String")]
-pub struct Hostname(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+pub struct DomainName(#[validate(regex(path = "validation::RFC_1123_SUBDOMAIN_REGEX"))] String);
+
+impl FromStr for DomainName {
+    type Err = validation::Errors;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        validation::is_rfc_1123_subdomain(value)?;
+        Ok(DomainName(value.to_owned()))
+    }
+}
 
-impl TryFrom<String> for Hostname {
+impl TryFrom<String> for DomainName {
     type Error = validation::Errors;
 
     fn try_from(value: String) -> Result<Self, Self::Error> {
-        validation::is_rfc_1123_subdomain(&value)?;
-        Ok(Hostname(value))
+        value.parse()
     }
 }
 
-impl From<Hostname> for String {
-    fn from(value: Hostname) -> Self {
+impl From<DomainName> for String {
+    fn from(value: DomainName) -> Self {
         value.0
     }
 }
 
-impl Display for Hostname {
+impl Display for DomainName {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         f.write_str(&self.0)
     }
 }
 
-impl Deref for Hostname {
+impl Deref for DomainName {
     type Target = str;
 
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
+#[derive(Debug, Snafu)]
+pub enum HostNameParseError {
+    #[snafu(display(
+        "the given hostname {hostname:?} is not a valid hostname, which needs to be either a domain name or IP address"
+    ))]
+    InvalidHostname { hostname: String },
+}
+
+/// A validated hostname, which is either a [`DomainName`] or [`IpAddr`].
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
+#[serde(try_from = "String", into = "String")]
+pub enum HostName {
+    IpAddress(IpAddr),
+    DomainName(DomainName),
+}
+
+impl JsonSchema for HostName {
+    fn schema_name() -> String {
+        "HostName".to_owned()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        String::json_schema(gen)
+    }
+}
+
+impl FromStr for HostName {
+    type Err = HostNameParseError;
+
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        if let Ok(ip) = value.parse::<IpAddr>() {
+            return Ok(HostName::IpAddress(ip));
+        }
+
+        if let Ok(domain_name) = value.parse() {
+            return Ok(HostName::DomainName(domain_name));
+        };
+
+        InvalidHostnameSnafu {
+            hostname: value.to_owned(),
+        }
+        .fail()
```

Comment:
This is dropping the context for _why_ validation failed.

Reply:
That's one way, but it loses the error chain. Another alternative would be to guess based on a cursory inspection.. something like:

```rs
if value.starts_with(char::is_ascii_digit) || value.contains(':') {
    Ok(HostName::IpAddress(value.parse().context(InvalidIpAddressSnafu { hostname: value })?))
} else {
    Ok(HostName::DomainName(value.parse().context(InvalidDomainNameSnafu { hostname: value })?))
}
```

Reply:
We could also do something like:

```rust
match value.parse::<IpAddr>() {
    Ok(ip_addr) => return Ok(Hostname::IpAddr(ip_addr)),
    Err(ip_err) => match value.parse::<DomainName>() {
        Ok(dn) => return Ok(Hostname::DomainName(dn)),
        Err(dn_err) => // return both errors
    }
}
```

Reply:
Yes that's effectively what @sbernauer suggested. There problem is that there's no way for an `Error` to have multiple `source()`s, so we'd have to pick one primary and render the other one as unawkwardly as we can.

Reply:
>  there's no way for an Error to have multiple source()s, so we'd have to pick one primary and render the other one as unawkwardly as we can.

That is indeed very true (and unfortunate). The cursed inspection is also kinda _meh_ tho. I don't know if it helps (I assume not), RFC 1123 states that the host name should first tried to be parsed as an IP address. So at least we do the correct thing RFC-wise.